[frontendActions] Add slide in new branch (closes #230) (#407)
VirtusLab/git-machete-intellij-plugin: docs/slide-in-use-cases.md

# Overview through Slide In use cases

This document is an overview through the use cases.
It should help to implement and understand the feature.

## Use cases

| UC ID | Local branch exists | Branch entry exists | Estimated frequency |
| :---: | :---: | :---: | :---: |
| 1 | false | false | ~ 75% |
| 2 | false | true  | < 1%  |
| 3 | true  | false | ~ 25% |
| 4 | true  | true  | < 1%  |

## Slide In - Logic

This specification covers all of the cases.

- When the **entry does not exist**, it must be added under the given branch.
- When the **entry exists under the given parent**, branch layout must not be affected.

> - When the **entry exists under other parent or is root**, it must be reattached under the given parent.

Reviewer: I think that some kind of question dialogbox will be nice in this case - this is something that (esp. beginning users) can overlook and this potentially can destroy branches tree.

Author: have you seen the slide in dialog? I think you might have omitted it.
![image](https://user-images.githubusercontent.com/19799111/87808613-506aa680-c85a-11ea-9867-97b27ce56291.png)

It clearly indicates what is to be slide in below what. 
Even though the newbie misses the **bold** parent he/she can easily undo things within (conveniently accessible) machete file. 

Adding the next dialog `Are you sure that you are sure?` is unnecessary.

Author: + actually the comment about dialog name applies here

Reviewer: Yes, I have seen this dialog, but IMHO i someone will try slide-in branch that is present under other parent this is not something fully intuitive that this branch will be reattached. I mean (esp. in large branch tree) this is simple to accidentally omit already added branch and trying to slide-in it again. But ok, I do not insist, maybe it's just my misconception that the user should be treated as a potential idiot xD

Author: I agree with your **first experience**.

Probably our different expectations (it was @PawelLipski suggestion to reattach) should be considered... :thinking: 
My idea is to add a checkbox (may be enabled by default) for the reattach (simply slide out and slide in without children otherwise). 

This is a candidate for another issue, WDYT?

Reviewer: Yes, checkbox is definitely good idea (maybe even better than messagebox) and yes, this can be candidate for other issue

Author: issue #439  created

> # Overview through Slide In use cases

Reviewer 2: Wow, that's become pretty short finally :hushed:

> It turned out that the sliding-in feature is more complex than it had appeared in the first place.

Reviewer 2: I don't think that it can be called `complex` anymore ;D

# Overview through Slide In use cases

It turned out that the sliding-in feature is more complex than it had appeared in the first place.
This document is an overview through the use cases.
It should help to implement and understand the feature.

## Slide In Branch Dialog

The picture below shows the **Slide In Branch Dialog**.

> This is a regular **Create New Branch Dialog** with a changed title.

Reviewer 2: I don't think this is up to date

![Slide In Branch Dialog](slide-in-branch-dialog.png "Slide In Branch Dialog")

Note the checkbox options:
- Checkout branch
- Overwrite existing branch

## Use cases

| UC ID | Local branch exists | Branch entry exists | Estimated frequency |
| :---: | :---: | :---: | :---: |
| 1 | false | false | ~ 75% |
| 2 | false | true  | < 1%  |
| 3 | true  | false | ~ 25% |
| 4 | true  | true  | < 1%  |

## Slide In - Logic

This specification covers all of the cases.

- When the **entry does not exist**, it must be added under the given branch.
- When the **entry exists under the given parent**, branch layout must not be affected.
- When the **entry exists under other parent or is root**, it must be reattached under the given parent.

**IMPORTANT NOTE:**

> Sliding in an entry under itself (directly and indirectly, as a deeper child node) must not be allowed.

Reviewer 2: (?)
```suggestion
Sliding in an entry under itself or any of its descendants in the branch layout must not be allowed.
```

### Zoom In On "Overwrite existing branch"

This section applies to **UC3** and **UC4**.

The same logic slide in logic adopts but overwriting itself requires a description.
The enabled overwrite checkbox is required for branches that already exist within a local repository.
Without it the overwrite (aka "New Branch Creation") is disabled (gray out "Create" button).

| Branch State | Overwrite Result |
| --- | :---: |
| **Working tree changed** (uncommitted changes) | Fail |
| Branch contains **some commits** | Fail |
| Branch contains **no commits** | Success |

![Checkout Failed Notification](checkout-failed.png "Checkout Failed Notification")

**IMPORTANT NOTE:**
The failure of the overwrite does not mean the failure of the slide in.

> Even though the notification like above may appear when the specified circumstances are met the branch slide will take place.

Reviewer 2: ```suggestion
The branch will still be slid in even if the overwrite fails.
```

Author: actually this section is irrelevant now

# Overview through Slide In use cases

It turned out that the sliding-in feature is more complex than it had appeared in the first place.

> This paper is an overview through the use cases.

Reviewer 2: `paper` is maybe too far-fetched a word for this doc ;)

It should help to implement and understand the feature.

## Slide In New Branch Dialog

The picture below shows the **Slide In New Branch Dialog**.
This is a regular **Create New Branch Dialog** with a changed title.

![Slide In New Branch Dialog](slide-in-new-branch-dialog.png "Slide In New Branch Dialog")

Note the checkbox options:
- Checkout branch
- Overwrite existing branch

## Use cases

| UC ID | Branch entry exists | Local branch exists | Estimated frequency |
| :---: | :---: | :---: | :---: |
| 1 | false | false | ~ 75% |
| 2 | false | true | ~ 25% |
| 3 | true | true | < 1% |
| 4 | true | false | < 1% |

## Slide In - Logic

- When the **entry does not exist**, it must be added (slid in) under the given branch.
- When the **entry exists under the given parent**, branch layout must not be affected.

> - When the **entry exists under other parent**, it must be first removed (slid out),

Reviewer 2: Also, an uncovered case here:

![image](https://user-images.githubusercontent.com/3383210/87126026-c8016a00-c28b-11ea-94c7-67524589a6eb.png)

**entry exists and is already root**, TBD what should be done (I think we could allow for attaching a root under a non-root branch, this might be useful e.g. for attaching `develop` under `master`)

Reviewer 2: Also, I don't think the error message should say "...of **new** branch master" since it has already existed

Reviewer 2: And also... I don't think more complex objects like `BranchLayout` should be passed to error messages like this one... I can address that on my follow-up PR

> and then added (slid in) under the given parent.

Reviewer 2: In terms of the user-facing naming: I think the user-facing dialog should be titled sth like `Add new branch or reattach existing branch under <parent-branch-name>`, the current `Slide In New Branch` is somewhat confusing wrt. those two distinct behaviors

Author: There are some factors here.
- we do not know whether the inserted branch name will be the name of an existing branch or a new one
- we shall care about consistency with the existing within IntelliJ `GitNewBranchDialog`
  - the most common case is creating a new branch 
  - it provides additional options (checkout, overwrite, (and in the near future) setTracking)

The dialog title can be changed easily to whatever we want. I am opting for `Slide In Branch Below <parent-name>`. 

Note the text field label `New branch name:` which references the `new`. Sadly, this one cannot be changed without reimplementation of the dialog.

## Slide In - New Branches

> This section applies to **UC1** and **UC4**.

Reviewer 2: Since the cases are naturally grouped by `Local branch exists` (so UC1&UC4 vs UC2&UC3) rather than by `Branch entry exists`, maybe reorder them (as, in swap the columns of the table)

Use case **UC1** seems to be the most common.
They are simply added in the way specified by the [slide in logic](#slide-in---logic).

> 

Reviewer 2: Hmm a word about UC4 missing?

## Slide In - With "Overwrite existing branch"

This section applies to **UC2** and **UC3**.

> **Overwriting a parent branch must not affect the branch layout**.

Reviewer 2: Hmm why would we want to "overwrite the (commit pointed by?) parent branch"? If I add Y under X... why would X be overwritten in the process?

Author: Probably already answered below (added to doc too).

The overwrite option is required for existing branches. Without it the dialog's "Create" button is disabled.

The other cases are described below.

| Branch State | Checkout | Branch Layout |
| --- | --- | --- |

> | **Working tree changed** | no checkout | branch entry **added*** under given parent |

Reviewer 2: Nit:
```suggestion
| **Working tree changed** | no checkout | branch entry **added**\* under given parent |
```
here and in the next rows

Reviewer 2: Also, not nit: what do you specifically mean by `Working tree changed`?

> | Branch contains **some commits** | no checkout | branch entry **added*** under given parent |

Reviewer 2: `Branch contains **some commits**`... as in, branch tip is not equal to its fork point? :thinking:

Reviewer 2: Does this row specifically refer to the case when this kind of message appears?

![image](https://user-images.githubusercontent.com/3383210/87127027-82de3780-c28d-11ea-846a-aa548504b746.png)

Reviewer 2: Although I have no idea based on what criteria git4idea decided that "some commits can be lost"

Reviewer 2: Steps to reproduce the above:
1. set up machete-sandbox from git-machete-blog-2
2. add `foo123` under `call-ws`
3. try sliding in under `hotfix/add-trigger`

Reviewer 2: (Note that as a last resort there's also a nuke-it-all solution: forbid reattaching existing branches completely and just limit ourselves to adding new branches)

Author: > Does this row specifically refer to the case when this kind of message appears?
> 
> ![image](https://user-images.githubusercontent.com/3383210/87127027-82de3780-c28d-11ea-846a-aa548504b746.png)

yup, this is the expected behavior 

This (and "Working tree changed" aka uncommitted changes) may cause failed checkout notification (but successfully slide in an existing branch entry!).

Author: My (probably wrong) thinking was that we can cover all cases 
\+ more by enhancing the existing new branch dialog by performing our slide in related operations (consistency, the same checkboxes options provided)
\- than creating a custom slide in dialog (unachievable consistency, more work to provide the checkbox options)

I would like to rethink things (on Monday) and consider **nuke-it-all solution**.

Author: One more thing about the present solution cause it may not have been clear. 

Sliding in of the existing branch is now possible only with **overwrite** option (without it the `Create` button is disabled for existing branches).

Author: Thinking out loud.

We could create a `Slide In Dialog` that takes the name (any valid) and slides in under the specified parent.
Then, for the NON-EXISTING BRANCHES ONLY show the `Create New Branch Dialog` using the provided name as the initial field value.
- a user may change the name within the second dialog (it would be rather hard to forbid the edition)
- we omit the problematic overwriting case
- TBD what to do in case of second dialog cancelation: do we want the whole action to be atomic or not

| Branch contains **no commits** | checkbox dependent | branch entry **added*** under given parent |

\* added in terms of [slide in logic](#slide-in---logic)

## Additional notes

> Slide In a branch under its own entry must not be allowed.

Reviewer 2: ```suggestion
Slide In a branch under itself must not be allowed.
```
